File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.cs
using System;
using System.Collections.Generic;

public class Solution {
    private int[] dx = {1, -1, 0, 0};
    private int[] dy = {0, 0, 1, -1};
    
    public int[] solution(string[,] places) {
        int[] answer = new int[places.GetLength(0)];
        
        for(int t = 0; t < places.GetLength(0); t++)
        {
            bool isOK = true;
            string[] curRoom = new string[5];
            for(int r = 0; r < 5; r++)
            {
                curRoom[r] = places[t, r];
            }
            for(int i = 0; i < 5; i++)
            {
                for(int j = 0; j < 5; j++)
                {
                    if(curRoom[i][j] == 'P')
                    {
                        if(!BFS(i, j, curRoom))
                        {
                            isOK = false;
                            break;
                        }
                    }
                }
                if(!isOK) break;
            }
            answer[t] = isOK ? 1 : 0;
        }
        return answer;
    }
    
    private bool BFS(int x, int y, string[] p)
    {
        var q = new Queue<(int row, int col)>();
        var visited = new bool[5, 5];
        
        q.Enqueue((x, y));
        visited[x, y] = true;
        
        while(q.Count > 0)
        {
            var cur = q.Dequeue();
            int cx = cur.row;
            int cy = cur.col;
            
            for(int i = 0; i < 4; i++)
            {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                
                if(nx < 0 || nx >= 5 || ny < 0 || ny >= 5 || visited[nx, ny])
                {
                    continue;
                }
                
                int d = Math.Abs(x - nx) + Math.Abs(y - ny);
                
                if(p[nx][ny] == 'P' && d <= 2)
                {
                    return false;
                }
                else if(p[nx][ny] == 'O' && d < 2)
                {
                    q.Enqueue((nx, ny));
                }
            }
        }
        return true;
    }
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

int dy[] = {-1, 0, 1, 0};
int dx[] = {0, 1, 0, -1};

int visited[5][5];

using namespace std;

bool bfs(int y, int x, vector<string> &place)
{
    memset(visited, -1, sizeof(visited));
    
    queue<pair<int, int>> q;
    
    q.push({y, x});
    visited[y][x] = 0;
    
    while(q.size())
    {
        tie(y, x) = q.front();
        q.pop();
        
        for(int i = 0; i < 4; i++)
        {
            int ny = y + dy[i];
            int nx = x + dx[i];
            
            if(ny < 0 || nx < 0 || ny >= 5 || nx >= 5)
                continue;
            if(visited[ny][nx] != -1)
                continue;
            if(visited[y][x] + 1 > 2)
                continue;
            
            visited[ny][nx] = visited[y][x] + 1;
            
            if(place[ny][nx] == 'X')
                continue;
            if(place[ny][nx] == 'P')
                return false;
            
            q.push({ny, nx});
        }
    }
    return true;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    for(auto place : places)
    {
        bool ok = true;
        
        for(int i = 0; i < 5; i++)
        {
            for(int j = 0; j < 5; j++)
            {
                if(place[i][j] == 'P')
                {
                    if(!bfs(i, j, place))
                        ok = false;
                }
            }
        }
        
        answer.push_back(ok ? 1 : 0);
    }
    return answer;
}